### ISSUE TITLE
Exporting Weights from Tensorflow Keras to C produces a different set of outputs, reduced accuracy in the model.

### ISSUE THREAD

**Issue by BeverlyCode** [keras-team/keras#17332]
For some reason when I export my weights from a Tensorflow Keras model, be it a simple Sequential FNN, and then load them into C to perform forward passes I get different output results, sometimes similar and often just not the same at all. It's like the accuracy has taken a substantial loss over the original predict() in Keras.

This is how I export my weights:
```
    for layer in model.layers:
        if layer.get_weights() != []:
            numpy.savetxt(layer.name + ".csv", layer.get_weights()[0].flatten(), delimiter=",")
            numpy.savetxt(layer.name + "_bias.csv", layer.get_weights()[1].flatten(), delimiter=",")
```
and
```
f.write(", /* weight */ " + str(layer.get_weights()[0].flatten()[i]))
f.write(", /* bias */ " + str(layer.get_weights()[1].flatten()[i]))
```
The first method writes a long E value and the second writes a smaller float value to the most significant decimal place. Although that should not cause a loss in precision over the first method? right?

I'm not sure why the saved weights would give different outputs, even with a 1 layer FNN that has 1 output,  a very simple network, all the weights are saved to float32 in C and then processed with float math functions such as tanhf() etc depending on the layer output.

I must be missing some understanding of what Tensorflow or Keras is doing behind the scenes? Something it's default configuration is doing to my network/weights that I need to disable or reproduce? Are the arrays of weights in reverse or something? Do I need to iterate them from left to right and not right to left on the inputs? What am I missing?

Here is a project that demonstrates this issue:
https://github.com/jcwml/neural_unitvector

In this project a dataset is exported from main.c to ensure the C program and Tensorflow Keras program both use data generated by the same random function, then Keras trains a set of weights for a simple FNN network to some number of hidden units and layers, and then those weights are exported and loaded into the C program where the forward pass is reconstructed in C code. There are 3 types of networks in this example that have had their forward passes reconstructed, each more complicated than the last, and yet, none of them produce the same outputs as Keras predict() had, it's not even a minor accuracy loss, it is substantially different, you can tell that the weights still represent what they where trained to but with just much worse prediction accuracy.

Another important fact to note is that when I train a network with multiple layers it would seem in many cases when I export the weights to C for a forward pass that the input data passed into the network eventually becomes 0's before it gets to the deeper layers, like vanishing gradients, this would imply that Tensorflow is using a different precision float format but the documentation says it uses float32 by default and the console printouts also claim to be using float32 in my application. My only thought is that maybe numpy is not saving them in the full float32 precision but when I looked into that apparently it does export full precision, and that exporting weights this way should be fine.

I am very confused.

**Comment by tilakrayal** [comment]
@BeverlyCode,
I was facing the issue/error while executing the given code. Kindly find the gist of it [here](https://colab.research.google.com/gist/tilakrayal/a48b6beef4df0723b6e191890be99b34/untitled808.ipynb).

**Comment by BeverlyCode** [comment]
> @BeverlyCode, I was facing the issue/error while executing the given code. Kindly find the gist of it [here](https://colab.research.google.com/gist/tilakrayal/a48b6beef4df0723b6e191890be99b34/untitled808.ipynb).

You can just not supply any command line parameters to the python file and it will execute fine.

**Comment by BeverlyCode** [comment]
Just a small update, from my research exporting the weights using both `numpy.savetxt(layer.name + ".csv", layer.get_weights()[0].flatten(), delimiter=",")` and `f.write("," + str(layer.get_weights()[0].flatten()[bc]))` on the latest Python versions and packages exports the weights to the exact decimal precision they are in.

But encase they didn't I also exported all the weight floats as bytes and then converted them back to floats in C to be extra-sure using this function on the export:
`def float_to_hex(f): return hex(struct.unpack('<I', struct.pack('<f', f))[0])`

It would seem that in even very basic FNN's in Tensorflow, there is something that Tensorflow is doing in the forward-pass that is not obvious, and I'm not sure how I would find-out what it is doing. I would assume that in Tensorflow the state of all things are disabled, such as layer regularisation, etc, unless explicitly turned on? Is it possible that Tensorflow actually turns some of this stuff on by default to improve default trained networks and that I just haven't realised this?

I'm clutching at straws here but Tensorflow is a large complicated SOTA framework and I'm not really sure where to start or how to go about debugging this.

**Comment by BeverlyCode** [comment]
Trying to figure this out independently still, I downloaded the Keras source code to try and find discrepancies, pdb was useless for this task for example, stepping though execution for a discrepancy is a lot less efficient when I can just check the code.

So I'm still in this process, Keras is a large system, it may take me a while. I am also going to look at [tensorflow-onnx](https://github.com/onnx/tensorflow-onnx) to see how they convert Keras models to ONNX for inference.

A comparison of the tanh function [activations.py#L379](https://github.com/keras-team/keras/blob/master/keras/activations.py#L379) shows very little difference between that and other implementations in terms of accuracy, a small C program enclosed demonstrates this:
```
#include <stdio.h>
#include <math.h>

const float e = 2.718281746F;
const float e2 = 7.389056206F;

static inline float mysinh(float f)
{
    return 0.5f * (powf(e, f) - powf(e, -f));
}

static inline float mycosh(float f)
{
    return (powf(e, f)/2.f) + (powf(e, -f)/2.f);
}

static inline float mytanh1(float f)
{
    return mysinh(f) / mycosh(f);
}

static inline float mytanh2(float f)
{
    const float pe = powf(e2, f);
    return (pe-1.f) / (pe+1.f);
}

static inline float kerasth(const float x) // keras tanh
{
    return ((expf(x) - expf(-x))/(expf(x) + expf(-x)));
}

int main()
{
    printf("mycosh():  %g\ncoshf():   %g\n", mycosh(45.f), coshf(45.f));
    printf("mytanh1(): %g\nmytanh2(): %g\ntanhf():   %g\ntanh():    %g\nkerasth(): %g\n", mytanh1(0.2f), mytanh2(0.2f), tanhf(0.2f), tanh(0.2f), kerasth(0.2f));
    printf("---\n");
    printf("mycosh():  %f\ncoshf():   %f\n", mycosh(45.f), coshf(45.f));
    printf("mytanh1(): %.9f\nmytanh2(): %.9f\ntanhf():   %.9f\ntanh():    %.9f\nkerasth(): %.9f\n", mytanh1(0.2f), mytanh2(0.2f), tanhf(0.2f), tanh(0.2f), kerasth(0.2f));
    return 0;
}
```
But I can get better results by dumping my sigmoid() for a simple scaled output, even though in Keras I define a sigmoid output on the final dense layer, when I import the weight in C I get better scaled outputs by replacing the sigmoid with an `output*0.03` for example. But that doesn't explain anything really, the outputs are still different, just detect the intended inputs more how it should, and it still doesn't explain why when I export the weights from an FNN with multiple layers that these weights quickly cause inputs to vanish past the first or second layer causing 0 ouputs.

It feels like either Keras is scaling the inputs or using some kind of regularisation between layers. Something is odd, but I am having a hard time workout out what.

**Comment by hertschuh** [comment]
Hi @BeverlyCode ,

It is really difficult to assess whether the code in `main.c` is equivalent to the Keras/Tensorflow code. Just the matrix multiplication, for instance, is difficult to analyze.

I do see a possible culprit though. In the example output of the exported keras model you added above, I see `"activation": "sigmoid"` on the last Dense layer. However, your C code only implements `tanh`. And in `fit.py`, the last layer has no activation right now. So there is a discrepancy here. Can you verify if that is the issue?

**Comment by BeverlyCode** [comment]
> Hi @BeverlyCode ,
> 
> It is really difficult to assess whether the code in `main.c` is equivalent to the Keras/Tensorflow code. Just the matrix multiplication, for instance, is difficult to analyze.
> 
> I do see a possible culprit though. In the example output of the exported keras model you added above, I see `"activation": "sigmoid"` on the last Dense layer. However, your C code only implements `tanh`. And in `fit.py`, the last layer has no activation right now. So there is a discrepancy here. Can you verify if that is the issue?

Nah that's just a model I dumped from a different network which I also output to C, some I use sigmoid some I use linear outputs. I have basically replicated many Dense FNN networks in C from Keras, and they all 'kind of work' but not really because the accuracy is lacking.

Nice try though, if only it was that simple. ;)

**Comment by hertschuh** [comment]
@BeverlyCode ,

I changed [line 208](https://github.com/jcwml/neural_unitvector/blob/main/fit.py#L208) to this:
```python
for weight in layer.get_weights()[0].transpose().flatten():
```
and tested in a roundabout way and got the same results.

So it's either a bug in the `norm_neural` code or the Python export code. It could just come from a misunderstanding of how `flatten` works.

Anyway, transposing before exporting fixes it.

Thanks.

**Comment by BeverlyCode** [comment]
You are completely right, you have solved it! Thank you so much!

I had to transpose the weight matrix before export, out of interest how did you discover this, I know that Tensorflow uses channels_last but the row/column major of the matrix was not a factor I had considered, partly I think because I did not even think of the weights as a matrix with columns or rows rather as a flat array.

I am wondering what you thought process was to have solved this so that I can improve my though processes.

But thank you so much. I hope that other people as confused as me find this thread and your answer. I should in future dump the array shape in future and if it is a matrix consider transposing I suppose.

**Comment by hertschuh** [comment]
Hi @BeverlyCode ,

My thought process was mostly a hunch that:
- it had nothing to do with Keras (an issue in something as simple as 2 Dense layers would have been found by now)
- it wasn't a precision issue, it was something more fundamental (numbers were way off)

Glancing at the C code, it looked correct. But one thing wasn't clear to me, which was whether the loops and index calculations were done correctly in the matrix multiplication to multiply rows with columns. Rather than trying to prove the code right or wrong, I thought I would transpose and give it a shot. And transposing the export was easier than modifying the code.

As far as the order of columns vs. rows is concerned, in this particular case, it's Numpy that's used (`get_weights` returns Numpy arrays and `flatten()` is a Numpy API, not Tensorflow). But as it turns out, Tensorflow uses the same order.

```python
import numpy as np
y = np.array([[1, 2], [3, 4]])
y.flatten()
```

```
array([1, 2, 3, 4])
```

```python
import tensorflow as tf
x = tf.constant([[1, 2], [3, 4]])
tf.reshape(x, (4,))
```

```
<tf.Tensor: shape=(4,), dtype=int32, numpy=array([1, 2, 3, 4], dtype=int32)>
```
